man/gd_alter_bit.3
.\" gd_alter_bit.3.  The gd_alter_bit man page.
.\"
.\" Copyright (C) 2008, 2009, 2010, 2012, 2013, 2014, 2016 D. V. Wiebe
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"
.\" This file is part of the GetData project.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.2 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, with no Front-Cover Texts, and with no Back-Cover
.\" Texts.  A copy of the license is included in the `COPYING.DOC' file
.\" as part of this distribution.
.\"
.TH gd_alter_bit 3 "22 November 2016" "Version 0.10.0" "GETDATA"
.SH NAME
gd_alter_bit, gd_alter_carray, gd_alter_clincom, gd_alter_const,
gd_alter_cpolynom, gd_alter_crecip, gd_alter_divide, gd_alter_indir,
gd_alter_lincom, gd_alter_linterp, gd_alter_multiply, gd_alter_phase,
gd_alter_polynom, gd_alter_raw, gd_alter_recip, gd_alter_sarray, gd_alter_sbit,
gd_alter_sindir
\(em modify a field in a dirfile
.SH SYNOPSIS
.B #include <getdata.h>
.HP
.nh
.ad l
.BI "int gd_alter_bit(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field ", int " bitnum ", int " numbits );
.HP
.BI "int gd_alter_carray(DIRFILE *" dirfile ", const char *" field_code ,
.BI "gd_type_t " const_type ", size_t " array_len );
.HP
.BI "int gd_alter_clincom(DIRFILE *" dirfile ", const char *" field_code ,
.BI "int " n_fields ", const char **" in_fields ", const double complex *" cm ,
.BI "const double complex *" cb );
.HP
.BI "int gd_alter_const(DIRFILE *" dirfile ", const char *" field_code ,
.BI "gd_type_t " const_type );
.HP
.BI "int gd_alter_cpolynom(DIRFILE *" dirfile ", const char *" field_code ,
.BI "int " poly_ord ", const char *" in_field ", const double complex *" ca );
.HP
.BI "int gd_alter_crecip(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field ", complex double " cdividend );
.HP
.BI "int gd_alter_divide(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field1 ", const char *" in_field2 );
.HP
.BI "int gd_alter_indir(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" index_field ", const char *" carray_field );
.HP
.BI "int gd_alter_lincom(DIRFILE *" dirfile ", const char *" field_code ,
.BI "int " n_fields ", const char **" in_fields ", const double *" m ,
.BI "const double *" b );
.HP
.BI "int gd_alter_linterp(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field ", const char *" table ", int " rename_table );
.HP
.BI "int gd_alter_mplex(DIRFILE *" dirfile ", const char *" field_name ,
.BI "const char *" in_field ", const char *" count_field ,
.BI "int " count_val ", int " period );
.HP
.BI "int gd_alter_multiply(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field1 ", const char *" in_field2 );
.HP
.BI "int gd_alter_phase(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field ", gd_shift_t " shift );
.HP
.BI "int gd_alter_polynom(DIRFILE *" dirfile ", const char *" field_code ,
.BI "int " poly_ord ", const char *" in_field ", const double *" ca );
.HP
.BI "int gd_alter_raw(DIRFILE *" dirfile ", const char *" field_code ,
.BI "gd_type_t " data_type ", unsigned int " spf ", int " recode );
.HP
.BI "int gd_alter_recip(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field ", double " dividend );
.HP
.BI "int gd_alter_sarray(DIRFILE *" dirfile ", const char *" field_code ,
.BI "size_t " array_len );
.HP
.BI "int gd_alter_sbit(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field ", int " bitnum ", int " numbits );
.HP
.BI "int gd_alter_sindir(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" index_field ", const char *" sarray_field );
.HP
.BI "int gd_alter_window(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field ", const char *" check_field ,
.BI "gd_windop_t " windop ", gd_triplet_t " threshold );
.hy
.ad n
.SH DESCRIPTION
These functions provide alternatives to using the
.BR gd_alter_entry (3)
function to modify a field of the indicated type in the dirfile specified by
.IR dirfile .
.PP
In all of these calls,
.I field_code
indicates the the field to be modified, which may be a regular field, or a
metafield specified by its full (slashed) field code, but should not contain a
representation suffix.  The meaning and valid
types of other arguments may be obtained from the
.BR get_entry (3)
and
.BR dirfile-format (5)
manual pages.  The
.B gd_shift_t
type is a signed 64-bit integer type.  The
.B gd_triplet_t
type is defined as:
.PP
.in +4n
.nf
.fam C
typedef union {
  gd_int64_t i;
  gd_uint64_t u;
  double r;
} gd_triplet_t;
.fam
.fi
.in
.PP
Which element of this
.B gd_triplet_t
union to set depends on the operator selected for the
.B WINDOW
field.  See
.BR gd_entry (3)
for details.

The
.BR gd_alter_clincom ()
and
.BR gd_alter_cpolynom ()
functions are identical to
.BR gd_alter_lincom ()
and
.BR gd_alter_polynom (),
except they take complex scalar parameters, instead of purely real values.  This
only matters for the input of
.I new
parameters; if the scalar parameters are not
changed (by passing NULL instead of a list of scalars), the functions can be
used interchangeably, regardless of whether the altered field has complex scalar
parameters or not.

If the corresponding parameters are to be changed, the
.BR gd_alter_lincom ()
and
.BR gd_alter_clincom ()
functions take pointers to three arrays of length
.I n_fields
containing the input field names
.RI ( in_fields ),
the gain factors
.RI ( m " or " cm ),
and the offset terms
.RI ( b " or " cb ).
Similarly,
.BR gd_alter_polynom ()
and
.BR gd_alter_cpolynom ()
take an array of length
.I poly_ord
+ 1 containing the polynomial co-efficients
.RI ( a " or " ca ).

Some field parameters have special values which indicate no change should be
made to the parameter.  These special values are:
.TP
.B NULL\fR:
any of the string parameters, also 
.IR m ", " b ", " a ", " cm ", " cb ", or " ca ;
.TP
.B 0\fR:
.IR spf ", " n_fields ", " numbits ", " cdividend ", " dividend ", or " array_len ;
.TP
.B -1\fR:
.IR bitnum " or " period ;
.TP
.B GD_NULL\fR:
.IR data_type " or " const_type ;
.TP
.B GD_WINDOP_UNK\fR:
.IR windop .
.PP
All field parameters introduced with this interface must contain literal
parameters.  Field parameters which are scalar fields cannot be introduced with
these functions.  To do that, use
.BR gd_alter_entry (3),
.BR gd_alter_spec (3)
or
.BR gd_malter_spec (3),
as appropriate.

If
.I rename_table
is non-zero, the look-up table referenced by the
.B LINTERP
field will be renamed to the path given by
.IR table .
If
.I recode
is non-zero, the binary file associated with the
.B RAW
field will be re-encoded to reflect the new field parameters.  In this case,
the field's I/O pointer will be reset to the beginning-of-frame.

If
.BR gd_alter_carray ()
is used to increase the length of a
.B CARRAY
field, the added elements will be uninitialised.  Use
.BR gd_put_carray_slice (3)
or equivalent to initialise them.  Similarly, increasing the length of a
.B SARRAY
with
.BR gd_alter_sarray ()
will set the added elements to the empty string ("").  Use
.BR gd_put_sarray_slice (3)
or equivalent to modify them.

See
.B NOTES
below for information on using
.BR gd_alter_clincom "(), " gd_alter_crecip (),
and 
.BR gd_alter_cpolynom ()
in the C89 GetData API.

.SH RETURN VALUE
On success, any of these functions returns zero.  On error, a negative-valued
error code is returned.  Possible error codes are:
.TP 8
.B GD_E_ACCMODE
The specified dirfile was opened read-only.
.TP
.B GD_E_ALLOC
The library was unable to allocate memory.
.TP
.B GD_E_BAD_CODE
The field specified by
.I field_code
was not found, or a supplied field code did not contain the appropriate prefix
or suffix.
.TP
.B GD_E_BAD_DIRFILE
The supplied dirfile was invalid.
.TP
.B GD_E_BAD_ENTRY
One or more of the field parameters specified was invalid.
.TP
.B GD_E_BAD_FIELD_TYPE
The field specified by
.I field_code
was of the wrong type for the function called.
.TP
.B GD_E_BAD_TYPE
The
.IR data_type " or " const_type
argument was invalid.
.TP
.B GD_E_IO
An I/O error occurred while translating the binary file associated with a
modified
.B RAW
field, or an I/O error occurred while attempting to rename a
.B LINTERP
table file.
.TP
.B GD_E_PROTECTED
The metadata of the fragment was protected from change.  Or, a request to
translate the binary file associated with a
.B RAW
field was attempted, but the data of the fragment was protected.
.TP
.B GD_E_UNKNOWN_ENCODING
The encoding scheme of the indicated format specification fragment is not known
to the library.  As a result, the library was unable to translate the binary
file be associated with a modified
.B RAW
field.
.TP
.B GD_E_UNSUPPORTED
The encoding scheme of the indicated format specification fragment does not
support translating the binary file associated with a modified
.B RAW
field.
.PP
The error code is also stored in the
.B DIRFILE
object and may be retrieved after this function returns by calling
.BR gd_error (3).
A descriptive error string for the error may be obtained by calling
.BR gd_error_string (3).

.SH NOTES
The C89 GetData API provides different prototypes for
.BR gd_alter_clincom "(), " gd_alter_cpolynom (),
and
.BR gd_alter_crecip ():
.PP
.nf
.B #define GD_C89_API
.br
.B #include <getdata.h>
.fi
.HP
.nh
.ad l
.BI "int gd_alter_clincom(DIRFILE *" dirfile ", const char *" field_code ,
.BI "int " n_fields ", const char **" in_fields ", const double *" cm ,
.BI "const double *" cb );
.HP
.BI "int gd_alter_cpolynom(DIRFILE *" dirfile ", const char *" field_code ,
.BI "int " poly_ord ", const char *" in_fields ", const double *" ca );
.HP
.BI "int gd_alter_crecip(DIRFILE *" dirfile ", const char *" field_code ,
.BI "const char *" in_field ", const double " cdividend\fR[2] );
.hy
.ad n
.PP
In this case, the array pointers passed as
.IR cm ", " cb
or
.IR ca
should have twice as many (purely real) elements, consisting of alternating
real and imaginary parts for the complex data.  That is, for example,
.IR ca [0]
should be the real part of the first co-efficient,
.IR ca [1]
the imaginary part of the first co-efficient,
.IR ca [2]
the real part of the second co-efficient,
.IR ca [3]
the imaginary part of the second co-efficient, and so on.  Similarly, the
.I cdividend
parameter becomes a double precision array of length two.
.PP
For
.BR gd_alter_clincom ()
and
.BR gd_alter_cpolynom (),
these are simply different (but equivalent) declarations of the C99 function
entry point.  For
.BR gd_alter_crecip (),
however, a different entry point is needed (since the
.I cdividend
parameter is passed by reference instead of by value).  In the interests of
portability, the C89 version of
.BR gd_alter_crecip ()
is always available, and may be accessed as
.BR gd_alter_crecip89 (),
with the C89 prototype, in both the C99 and C89 APIs.  Passing NULL as
.I cdividend
is equivalent to specifying a dividend of zero: it indicates no change to the
dividend parameter.

.SH SEE ALSO
.BR gd_alter_entry (3),
.BR gd_alter_spec (3),
.BR gd_error (3),
.BR gd_error_string (3),
.BR gd_malter_spec (3),
.BR gd_metaflush (3),
.BR gd_open (3),
.BR gd_put_carray_slice (3),
.BR gd_put_sarray_slice (3),
.BR dirfile-format (5)


NEWS
New in version 0.10.0a:

  Dirfile Changes:

  * Dirfile Standards Version 10 has been released.  It adds three new
    field types: SARRAY, which is an array of STRING scalars (like a
    CARRAY, but with STRINGs instead of CONSTs), and INDIR and SINDIR,
    which are vector fields provide indexed look-ups from CARRAY (INDIR) or
    SARRAY (SINDIR) scalar arrays.  It also adds field code namespaces,
    which can be specified with the new /NAMESPACE directive, or else as
    part of an /INCLUDE directive.

  * Some notes about namespaces:

    - Namespaces are separated from field names by a dot (.):
  
          namespace.name
        
      and can be nested arbitrarily deep, separating namespaces with
      intermediate dots:

          namespace.subspace.subsubspace.name

      Namespaces were created to provide an alternative to the prefix and
      suffix added to the /INCLUDE driective in Standards Version 9, which
      have some unfortunate side-effects due to their modifying field names
      directly.  Because namespaces nest and are syntactically separate
      from field names, they do a better job of encapsulation.

    - If the namespace of a field is null (""), which is the default, then
      the dot separing the namespace from the field name may be omitted.

    - An /INCLUDE statement can specify a namespace which becomes the
      included fragment's "root namespace".  The root namespace of the base
      (top-level) format file is always the nullspace (""), and cannot be
      changed.

    - In additon to root namespaces, there is also a "current namespace".
      At the top of a fragment, the current namespace is set to the root
      namespace.  A /NAMESPACE directive can be used to change the current
      namespace to a subspace under the root namespace.  That is, if the
      root namespace is "root", then the directive:

          /NAMESPACE subspace

       changes the current namespace to "root.subspace" regardless of what
       the current namespace was before.  To change the current namespace
       back to the root namespace, use the null token ("") with the
       /NAMESPACE directive:

          /NAMESPACE ""

    - If no namespace is specified in an /INCLUDE line, then the current
      namespace becomes the root namespace of the included fragment.

    - Subnamespaces under the root namespace may also be specified directly
      in the name part of a field specification.  As a result, it is never
      necessary to use the /NAMESPACE directive.

    - Every field code and name in a fragment implicitly gains either the
      fragment's root namespace or the current namespace.  If the field
      code starts with an initial dot, then the root namespace is prepended
      to it, otherwise the current namespace is prepended to it.  Because
      the current namespace is always a subspace of the root namespace,
      this means that metadata in a given fragment is never able to access
      other fields outside its own root namespace.

    - The exception to the above is the implicit INDEX vector, which
      ignores all namespaces attached to it, either implicitly, through the
      current namespace, or explicitly when specified in the metadata.
      This contrasts with the behaviour of the INDEX field in the presence
      of affixes, where it can appears or disappears based on the effects
      of the affixes creating or modifying the specific field name "INDEX".

  * Because syntactially the dot (.) now performs two functions, namely
    both separating namespaces from each other and field names, but also
    separating a field code from a representation suffix, there exists
    ambiguity in the syntax.  To resolve the ambiguity, a new representa-
    tion suffix, .z has been added which does nothing.  As an example, the
    field code:

      name.r

    is interpreted as the real part of the field named "name", asuming such
    a field exists.  To indicate the field named "r" in the namespace
    "name", the field code:

      name.r.z

    must be used.  Note that this ambiguity only exists in a Dirfile where
    both "name" and "name.r" are valid field names.  If the field named
    "name" doesn't exist, then the first field code, "name.r" is unambig-
    uously interpreted as the field "r" in the namespace "name".

  * A note on the SINDIR field: Unlike every other vector field, this
    field produces character string data at the sample rate of it's input
    vector, which may be surprising, in certain instances.

  * GetData has supported FLAC compression since 0.9.0.  Standards Version
    10 now adds "flac" to the list of pre-defined encodings.

  Library Changes:

  * The function gd_array_len() no longer silently ignores a representation
    suffix in the field_code provided.  In most cases, passing a
    representation suffix will now cause this function to fail with a
    GD_E_BAD_CODE error.

  * A couple of unnecessary malloc's have been removed from the field
    code search, reducing the chance of encountering a GD_E_ALLOC error.
    Notably, the functions gd_array_len(), gd_bof(), gd_entry_type(),
    gd_fragment_index(), and gd_spf() will no longer produce this error at
    all.

  * BUG FIX: The gd_include() family of functions now correcrtly clean up
    after encountering an error.  Previously, encountering a syntax error
    would result in these functions erroneously adding fields specified
    before the syntax error to the DIRFILE, with a bogus fragment index.
    Similarly, when these functions return GD_E_REFERENCE, they no longer
    add the included fragment (excluding the bad /REFERENCE directive) to
    the DIRFILE without telling the caller about it.

  * BUG FIX: gd_alter_protection() wasn't marking affected fragments as
    modified, meaning the change in protection level would be lost upon
    close unless other metadata changes were made to the fragment as well.


bindings/matlab/gd_alter_bit.m
function gd_alter_bit(D, field_code, in_fields, bitnum, numbits)
% GD_ALTER_BIT  Modify the metadata of a BIT field
%
%   GD_ALTER_BIT(DIRFILE,NAME,INPUT,BITNUM,NUMBITS)
%             modifies the metadata of the BIT field called NAME in the
%             dirfile specified by DIRFILE.  The input field is set to INPUT,
%             the first bit to BITNUM, and the number of bits to NUMBITS.
%
%   The DIRFILE object should have previously been created with GD_OPEN.
%
%   See the documentation on the C API function gd_alter_bit(3) in section 3
%   of the UNIX manual for more details.
%
%   See also GD_ALTER_ENTRY, GD_OPEN

  GD = getdata_constants();
  gd_alter_entry(D, field_code, struct('field_type', GD.BIT_ENTRY, ...
  'in_fields', {in_fields}, 'bitnum', bitnum, 'numbits', numbits), 0);
end

% Copyright (C) 2013 D. V. Wiebe
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% This file is part of the GetData project.
%
% GetData is free software; you can redistribute it and/or modify it under
% the terms of the GNU Lesser General Public License as published by the
% Free Software Foundation; either version 2.1 of the License, or (at your
% option) any later version.
%
% GetData is distributed in the hope that it will be useful, but WITHOUT
% ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
% FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
% License for more details.
%
% You should have received a copy of the GNU Lesser General Public License
% along with GetData; if not, write to the Free Software Foundation, Inc.,
% 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
